### PR TITLE
#413: Added additional constructors to HmRqHeader

### PR DESCRIPTION
**Pull request by bdragan**
#413: Added additional constructors to HmRqHeader.

### CODE REVIEW

**Comment by davvd** [comment]
@bdragan Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@pinaf please review, thanks

File: src/main/java/org/takes/facets/hamcrest/HmRqHeader.java [hunk 1]
```diff
@@ -86,38 +101,124 @@ public void describeTo(final Description description) {
...
+        public boolean matches(final RqHeaders headers) throws IOException {
             boolean result = false;
-            final RqHeaders headers = new RqHeaders.Base(item);
-            for (final String name: headers.names()) {
-                if (this.matchHeader(name, headers.header(name))) {
-                    result = true;
-                    break;
+            outer:
+            for (final String name : headers.names()) {
+                for (final String value: headers.header(name)) {
```
Comment (pinaf) [inline src/main/java/org/takes/facets/hamcrest/HmRqHeader.java]: @bdragan let's add
```java
if (result) {
    break;
}
```

File: src/main/java/org/takes/facets/hamcrest/HmRqHeader.java [hunk 2]
```diff
@@ -86,38 +101,124 @@ public void describeTo(final Description description) {
...
+        @Override
+        @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+        public boolean matches(final RqHeaders headers) throws IOException {
             boolean result = false;
-            final RqHeaders headers = new RqHeaders.Base(item);
-            for (final String name: headers.names()) {
-                if (this.matchHeader(name, headers.header(name))) {
-                    result = true;
-                    break;
+            outer:
```
Comment (pinaf) [inline src/main/java/org/takes/facets/hamcrest/HmRqHeader.java]: @bdragan let's remove this label

File: src/main/java/org/takes/facets/hamcrest/HmRqHeader.java [hunk 3]
```diff
@@ -86,38 +101,124 @@ public void describeTo(final Description description) {
...
+            for (final String name : headers.names()) {
+                for (final String value: headers.header(name)) {
+                    final Map.Entry<String, String> entry =
+                        new EntryImpl<String, String>(
+                            name.trim().toLowerCase(Locale.ENGLISH),
+                            value.trim()
+                        );
+                    if (this.matcher.matches(entry)) {
+                        result = true;
+                        break outer;
```
Comment (pinaf) [inline src/main/java/org/takes/facets/hamcrest/HmRqHeader.java]: @bdragan let's do a simple `break` here

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 4]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
                 Matchers.not(
-                    Matchers.hasEntry("host", "fake.org")
+                    new EntryMatcher<String, String>("host", "fake.org")
                 )
             )
         );
     }
+
+    /**
+     * HmRqHeader can test header name and value available.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan can test `whether` header name...

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 5]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
+    @Test
+    public void testsHeaderNameAndValueAvailable() throws Exception {
+        MatcherAssert.assertThat(
+            new RqWithHeader(new RqFake(), "header1: value1"),
+            new HmRqHeader("header1", "value1")
+        );
+    }
+
+    /**
+     * HmRqHeader can test header name and value not available.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan can test `whether header name is available and value is not available.`

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 6]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
+    @Test
+    public void testsHeaderNameAndValueNotAvailable() throws Exception {
+        MatcherAssert.assertThat(
+            new RqWithHeader(new RqFake(), "header2: value2"),
+            Matchers.not(new HmRqHeader("header2", "value21"))
+        );
+    }
+
+    /**
+     * HmRqHeader can test headers available.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan can test `whether multiple headers are available.`

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 7]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void testsHeadersAvailable() throws Exception {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan testsMultipleHeadersAvailable`

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 8]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
+            new HmRqHeader("header1", "value1")
+        );
+    }
+
+    /**
+     * HmRqHeader can test header name and value not available.
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void testsHeaderNameAndValueNotAvailable() throws Exception {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan `testsValueNotAvailable`

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 9]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
+            new RqWithHeaders(
+                new RqFake(),
+                "header3: value31", "header3: value32"
+            ),
+            new HmRqHeader("header3", Matchers.<String>iterableWithSize(2))
+        );
+    }
+
+    /**
+     * HmRqHeader can test headers not available.
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan can test `whether a header is not available`

File: src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java [hunk 10]
```diff
@@ -74,9 +80,60 @@ public void testsHeaderNotAvailable() throws Exception {
...
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void testsHeadersNotAvailable() throws Exception {
```
Comment (pinaf) [inline src/test/java/org/takes/facets/hamcrest/HmRqHeaderTest.java]: @bdragan `testsHeaderNotAvailable`

**Comment by pinaf** [comment]
@bdragan nice job

**Comment by pinaf** [comment]
@bdragan 10 comments

**Comment by bdragan** [comment]
@pinaf Thanks. Please take a look at the new commit.

**Comment by pinaf** [comment]
@rultor merge

**Comment by pinaf** [comment]
@bdragan thanks!

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova please, review this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@davvd Quality is good here.

**Comment by davvd** [comment]
> @davvd Quality is good here.

@elenavolokhova thank you, good quality is always good ;)

**Comment by davvd** [comment]
@pinaf just added 10 mins to @elenavolokhova (for QA), payment ID is `70104518`. Thanks for your contribution, **29 mins** was added to your account, payment ID is `AP-70323889J3006360N`, 47 hours and 5 mins spent total. extra minutes for review comments (c=14). +29 added to your rating, at the moment it is: [+4326](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**